File: bridges/nodejs/src/sdk/network.ts
import dns from 'node:dns'

import type { AxiosInstance } from 'axios'
import axios from 'axios'

import { LEON_VERSION, NODEJS_BRIDGE_VERSION } from '@bridge/constants'

interface NetworkOptions {
  /** `baseURL` will be prepended to `url`. It can be convenient to set `baseURL` for an instance of `Network` to pass relative URLs. */
  baseURL?: string
}

interface NetworkRequestOptions {
  /** Server URL that will be used for the request. */
  url: string

  /** Request method to be used when making the request. */
  method: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE'

  /** Data to be sent as the request body. */
  data?: unknown

  /** Custom headers to be sent. */
  headers?: Record<string, string>

  /** Optional files for multipart/form-data requests (parity with Python SDK). */
  files?: Record<string, unknown>

  /** Whether to send JSON body (true by default). If false, send form data. */
  useJson?: boolean

  /** Response type (defaults to 'json'). Use 'arraybuffer' for binary data like audio/video files. */
  responseType?:
    | 'json'
    | 'text'
    | 'arraybuffer'
    | 'blob'
    | 'document'
    | 'stream'
}

interface NetworkResponse<ResponseData> {
  /** Data provided by the server. */
  data: ResponseData

  /** HTTP status code from the server response. */
  statusCode: number

  /** Options that was provided for the request. */
  options: NetworkRequestOptions & NetworkOptions
}

export class NetworkError<ResponseErrorData = unknown> extends Error {
  public readonly response: NetworkResponse<ResponseErrorData>

  constructor(response: NetworkResponse<ResponseErrorData>) {
    super(`[NetworkError]: ${response.statusCode} ${response.data}`)
    this.response = response
    Object.setPrototypeOf(this, NetworkError.prototype)
  }
}

export class Network {
  private readonly options: NetworkOptions
  private axios: AxiosInstance

  constructor(options: NetworkOptions = {}) {
    this.options = options
    this.axios = axios.create({
      baseURL: this.options.baseURL
    })
  }

  /**
   * Send HTTP request
   * @param options Request options
   * @example request({ url: '/send', method: 'POST', data: { message: 'Hi' } })
   */
  public async request<ResponseData = unknown, ResponseErrorData = unknown>(
    options: NetworkRequestOptions
  ): Promise<NetworkResponse<ResponseData>> {
    try {
      const response = await this.axios.request({
        url: options.url,
        method: options.method.toLowerCase(),
        // For parity, we accept any data type here (including FormData)
        data: options.data as never,
        responseType: options.responseType,
        headers: {
          'User-Agent': `Leon Personal Assistant ${LEON_VERSION} - Node.js Bridge ${NODEJS_BRIDGE_VERSION}`,
          ...options.headers
        }
      })

      let data = {} as ResponseData
      // For binary response types, return data as-is
      if (
        options.responseType === 'arraybuffer' ||
        options.responseType === 'blob' ||
        options.responseType === 'stream'
      ) {
        data = response.data as ResponseData
      } else {
        // For text/json responses, try to parse as JSON
        try {
          if (typeof response.data === 'string') {
            data = JSON.parse(response.data)
          } else {
            data = response.data as ResponseData
          }
        } catch {
          data = response.data as ResponseData
        }
      }

      return {
        data,
        statusCode: response.status,
        options: {
          ...this.options,
          ...options
        }
      }
    } catch (error) {
      let statusCode = 500
      let dataRawText = ''

      if (axios.isAxiosError(error)) {
        dataRawText = error?.response?.data ?? ''
        statusCode = error?.response?.status ?? 500
      }

      let data: ResponseErrorData
      try {
        data = JSON.parse(dataRawText)
      } catch {
        data = dataRawText as ResponseErrorData
      }

      throw new NetworkError<ResponseErrorData>({
        data,
        statusCode,
        options: {
          ...this.options,
          ...options
        }
      })
    }
  }

  /**
   * Check if error is a network error
   * @param error Error to check
   * @example isNetworkError(error) // false
   */
  public isNetworkError<ResponseErrorData = unknown>(
    error: unknown
  ): error is NetworkError<ResponseErrorData> {
    return error instanceof NetworkError
  }

  /**
   * Verify whether there is an Internet connectivity
   * @example isNetworkAvailable() // true
   */
  public async isNetworkAvailable(): Promise<boolean> {
    try {
      await dns.promises.resolve('getleon.ai')

      return true
    } catch {
      return false
    }
  }
}

File: bridges/python/src/sdk/network.py
import requests
import socket
from typing import Any, Dict, TypedDict, Union, Literal, Optional

from ..constants import LEON_VERSION, PYTHON_BRIDGE_VERSION


class NetworkOptions(TypedDict, total=False):
    base_url: Optional[str]


class NetworkResponse(TypedDict):
    data: Any
    status_code: int
    options: Dict[str, Any]


class NetworkError(Exception):
    def __init__(self, response: NetworkResponse) -> None:
        self.response = response
        super().__init__(f"[NetworkError]: {response['status_code']} {response['data']}")


class NetworkRequestOptions(TypedDict, total=False):
    url: str
    method: Union[Literal['GET'], Literal['POST'], Literal['PUT'], Literal['PATCH'], Literal['DELETE']]
    data: Dict[str, Any]
    headers: Dict[str, str]
    files: Dict[str, Any]
    use_json: bool
    response_type: Optional[Union[Literal['json'], Literal['text'], Literal['arraybuffer'], Literal['bytes']]]

class Network:
    def __init__(self, options: NetworkOptions = {'base_url': None}) -> None:
        self.options = options

    def request(self, options: NetworkRequestOptions) -> NetworkResponse:
        try:
            url = options['url']

            if self.options['base_url'] is not None:
                url = (self.options['base_url'] or '') + url

            method = options['method']
            data = options.get('data', {})
            headers = options.get('headers', {})
            files = options.get('files')
            use_json = options.get('use_json', True)
            response_type = options.get('response_type', 'json')

            request_kwargs: Dict[str, Any] = {
                'headers': {
                    'User-Agent': f"Leon Personal Assistant {LEON_VERSION} - Python Bridge {PYTHON_BRIDGE_VERSION}",
                    **headers
                }
            }

            # If files are provided or JSON is explicitly disabled, send form data
            if files or not use_json:
                request_kwargs['data'] = data
                if files:
                    request_kwargs['files'] = files
            else:
                request_kwargs['json'] = data

            response = requests.request(
                method,
                url,
                **request_kwargs
            )

            parsed_data: Any
            if response_type in ['arraybuffer', 'bytes']:
                parsed_data = response.content
            else:
                try:
                    parsed_data = response.json()
                except Exception:
                    parsed_data = response.text

            network_response: NetworkResponse = {
                'data': parsed_data,
                'status_code': response.status_code,
                'options': {**self.options, **options}
            }

            if response.ok:
                return network_response
            else:
                raise NetworkError(network_response)
        except requests.exceptions.RequestException as error:
            status_code = 500
            raw_data: Any = ''

            if error.response is not None:
                status_code = error.response.status_code
                try:
                    raw_data = error.response.json()
                except Exception:
                    raw_data = error.response.text

            raise NetworkError({
                'data': raw_data,
                'status_code': status_code,
                'options': {**self.options, **options}
            }) from error

    def is_network_error(self, error: Exception) -> bool:
        return isinstance(error, NetworkError)

    def is_network_available(self) -> bool:
        try:
            socket.gethostbyname('getleon.ai')

            return True
        except socket.error:
            return False
